interactions/api/voice/audio.py

```diff
@@ -5,19 +5,70 @@
 import time
 from abc import ABC, abstractmethod
 from pathlib import Path
-from typing import Union, Optional
+from typing import Union, Optional, TYPE_CHECKING
 
-__all__ = (
-    "AudioBuffer",
-    "BaseAudio",
-    "Audio",
-    "AudioVolume",
-)
+__all__ = ("AudioBuffer", "BaseAudio", "Audio", "AudioVolume", "RawInputAudio")
 
 from interactions.client.const import get_logger
 from interactions.api.voice.opus import Encoder
 from interactions.client.utils import FastJson
 
+if TYPE_CHECKING:
+    from interactions.api.voice.recorder import Recorder
+
+
+class RawInputAudio:
+    decoded: bytes
+    """The decoded audio"""
+    pcm: bytes
+    """The raw PCM audio"""
+    sequence: int
+    """The audio sequence"""
+    audio_timestamp: int
+    """The current timestamp for this audio"""
+    timestamp_ns: float
+    """The time this audio was received, in nanoseconds"""
+    timestamp: float
+    """The time this audio was received, in seconds"""
+    ssrc: int
+    """The source of this audio"""
+    _recoder: "Recorder"
+    """A reference to the audio recorder managing this object"""
+
+    def __init__(self, recorder: "Recorder", data: bytes) -> None:
+        self.decoded: bytes = b""
+        self._recorder = recorder
+        self.timestamp_ns = time.monotonic_ns()
+        self.timestamp = self.timestamp_ns / 1e9
+        self.pcm = b""
+
+        self.ingest(data)
+
+    def ingest(self, data: bytes) -> bytes | None:
+        data = bytearray(data)
+        header = data[:12]
+
+        decrypted: bytes = self._recorder.decrypt(header, data[12:])
+        self.ssrc = int.from_bytes(header[8:12], byteorder="big")
+        self.sequence = int.from_bytes(header[2:4], byteorder="big")
+        self.audio_timestamp = int.from_bytes(header[4:8], byteorder="big")
+
+        if not self._recorder.recording_whitelist or self.user_id in self._recorder.recording_whitelist:
+            # noinspection PyProtectedMember
+            if decrypted[0] == 0xBE and decrypted[1] == 0xDE:
+                # rtp header extension, remove it
+                header_ext_length = int.from_bytes(decrypted[2:4], byteorder="big")
+                decrypted = decrypted[4 + 4 * header_ext_length :]
+            self.decoded = self._recorder.get_decoder(self.ssrc).decode(decrypted)
+            return self.decoded
+
+    @property
+    def user_id(self) -> Optional[int]:
+        """The ID of the user who made this audio."""
+        while not self._recorder.state.ws.user_ssrc_map.get(self.ssrc):
+            time.sleep(0.05)
```

Comment:
Isn't this blocking?

Reply:
Yes, if it were to be called from an async function, which it never is. RawAudio is only ever used from within a thread

Reply:
Oh ok